xmtp_mls/src/storage/encrypted_store/mod.rs

```diff
@@ -393,23 +396,23 @@ where
         tracing::debug!("Transaction beginning");
         {
             let connection = self.conn_ref();
-            let mut connection = connection.inner_mut_ref();
+            let mut connection = connection.read_mut_ref();
```

Comment:
Transactions are directly tied to their connection, so there may be trouble by using the read connection interchangeably with the write connection. The write transaction will have started `COMMIT TRANSACTION` but a `BEGIN TRANSACTION` pragma will never have happened because it occurred on the previous read reference

Reply:
This does pose another problem though, if we use `begin_transaction` and then release the mutex lock, I think we can still run into a db lock issue

- begin transaction with write conn
- do a write
- tx releases lock
- yield b/c tx is async and does a network call
- begin another tx with write lock somewhere else
    - completely legal because we relinquisheed the mutex after begin_transaction
- db lock even though we're using a mutex, because although the mutex exists in libxmtp, the transaction is open in SQLite with the same connection, which conflicts with another transaction on the same database



If we instead ensure the mutex is held for the entire transaction, we'll just run into a deadlock cycle waiting for access to the RwLock

Reply:
Yeah, you're right. It does deadlock